commit ae495e8bd485f48ecdb7e53d7e98771220f31997
Author: Zoltan Varga <vargaz@gmail.com>
Date:   Wed May 6 04:56:52 2015 -0400

    [runtime] Fix support for callvirt delegates with a vtype first argument. Fixes #29665.

diff --git a/mcs/class/corlib/Test/System/DelegateTest.cs b/mcs/class/corlib/Test/System/DelegateTest.cs
index 843f48c..2458b98 100644
--- a/mcs/class/corlib/Test/System/DelegateTest.cs
+++ b/mcs/class/corlib/Test/System/DelegateTest.cs
@@ -1069,6 +1069,27 @@ namespace MonoTests.System
 			action_int (42);
 		}
 
+		struct FooStruct {
+			public int i, j, k, l;
+
+			public int GetProp (int a, int b, int c, int d) {
+				return i;
+			}
+		}
+
+		delegate int ByRefDelegate (ref FooStruct s, int a, int b, int c, int d);
+
+#if MONOTOUCH
+		[Category ("NotWorking")]
+#endif
+		[Test]
+		public void CallVirtVType ()
+		{
+			var action = (ByRefDelegate)Delegate.CreateDelegate (typeof (ByRefDelegate), null, typeof (FooStruct).GetMethod ("GetProp"));
+			var s = new FooStruct () { i = 42 };
+			Assert.AreEqual (42, action (ref s, 1, 2, 3, 4));
+		}
+
 		class Foo {
 
 			public void Bar ()
diff --git a/mono/metadata/marshal.c b/mono/metadata/marshal.c
index ba6494c..5dfefae 100644
--- a/mono/metadata/marshal.c
+++ b/mono/metadata/marshal.c
@@ -3231,11 +3231,18 @@ mono_marshal_get_delegate_invoke_internal (MonoMethod *method, gboolean callvirt
 
 	if (callvirt) {
 		if (!closed_over_null) {
-			mono_mb_emit_ldarg (mb, 1);
-			mono_mb_emit_op (mb, CEE_CASTCLASS, target_class);
-			for (i = 1; i < sig->param_count; ++i)
-				mono_mb_emit_ldarg (mb, i + 1);
-			mono_mb_emit_op (mb, CEE_CALLVIRT, target_method);
+			if (target_class->valuetype) {
+				mono_mb_emit_ldarg (mb, 1);
+				for (i = 1; i < sig->param_count; ++i)
+					mono_mb_emit_ldarg (mb, i + 1);
+				mono_mb_emit_op (mb, CEE_CALL, target_method);
+			} else {
+				mono_mb_emit_ldarg (mb, 1);
+				mono_mb_emit_op (mb, CEE_CASTCLASS, target_class);
+				for (i = 1; i < sig->param_count; ++i)
+					mono_mb_emit_ldarg (mb, i + 1);
+				mono_mb_emit_op (mb, CEE_CALLVIRT, target_method);
+			}
 		} else {
 			mono_mb_emit_byte (mb, CEE_LDNULL);
 			for (i = 0; i < sig->param_count; ++i)
